### ISSUE TITLE
Select option empty value regression in 15.0.0-rc.1

### ISSUE THREAD

**Issue by chicoxyzzy** [facebook/react#6219]
React 0.14.7: `<option value="">empty</option>` emits empty string on change
http://jsbin.com/pisita/10/edit?html,js,console,output
React 15.0.0-rc.1`<option value="">empty</option>` emits string with value "empty" on change
http://jsbin.com/bemaze/12/edit?html,js,console,output
This is because React 15.0.0-rc.1 cuts off `value=""` from option tag

**Comment by chicoxyzzy** [comment]
I don't know why jsbin doesn't save html...

**Comment by chicoxyzzy** [comment]
html for 0.14.7
```html
<script src="https://fb.me/react-with-addons-0.14.7.min.js"></script>
<script src="https://fb.me/react-dom-0.14.7.min.js"></script>
<div id="app"></div>
```

html for 15.0.0-rc.1
```html
<script src="https://fb.me/react-with-addons-15.0.0-rc.1.min.js"></script>
<script src="https://fb.me/react-dom-15.0.0-rc.1.min.js"></script>
<div id="app"></div>
```

**Comment by ngthorg** [comment]
you try to write like this look
```
var Select = React.createClass({
  displayName: 'Select',
  handleOnChange: function(e) {
    console.log(e.target.value);
  },
  render: function() {
    return (
      <select onChange={this.handleOnChange}>
        <option value="">empty</option>
        <option value="filled">filled</option>
      </select>
    );
  }
});
```

**Comment by chicoxyzzy** [comment]
I want to fix this. Can you point me to right place? I believe this is ReactDOM issue because Virtual DOM node looks ok.

**Comment by jimfb** [comment]
Yeah, I spent a few minutes looking for the PR that I suspect is responsible for this, but I couldn't find it.  I **suspect** I remember seeing the PR that probably introduced this - I think it was intended to use the child text value if the element's `value` prop was undefined, but it appears to be capturing the empty string as well.  Of course, I could be totally wrong, I'm speculating based on a distant memory of a PR, which is always a dangerous thing to do.

**Comment by jimfb** [comment]
Without having actually dug into the code, my intuition is to look a https://github.com/facebook/react/blob/master/src/renderers/dom/client/wrappers/ReactDOMSelect.js and https://github.com/facebook/react/blob/master/src/renderers/dom/client/wrappers/ReactDOMOption.js

But don't rely on those tips too heavily, I don't want to send you off on a wild goose chase.

**Comment by everdimension** [comment]
I believe this commit introduces the bug (changes to file `DOMPropertyOperations.js`): https://github.com/facebook/react/commit/77a137adf917c0ab87a6c9b6fec58938b14c38ce#diff-f04a397aa63ac69c28dab46e9863bb90L171

This code is responsible for setting attributes to the nodes being created.

It's not a straightforward fix: 
— the older version checked the `mustUseAttribute` (which has now been removed) property and was pretty much setting the attributes in any case;
— current version checks for `mustUseProperty` property, but only sets the `value` attribute if it is not falsy.

It's questionable whether the current behavior is correct. Basically it checks whether the attribute is classified as `HAS_SIDE_EFFECTS` which is only true (at least for now) for the `value` attribute.
And then it does this quirky comparison `('' + node[propName]) !== ('' + value)` to set the node property.

The `option` element is in a sort of unique position: If no `value` attribute value is provided, then the tag's content is used as value. If `value` attribute exists, even if it is empty, then the attribute is used as the value.

But, for example, the `input[type=checkbox]` (same for `radio`) also behaves similar in these cases:
* it returns `on` as its value if no `value` attribute is provided
* it returns the attribute's value if `value` exists, even if empty

The thing is, when the abovementioned check is done, before `value` attribute is applied, `node.value` for checkbox returns "on" and that's why the attribute is set as expected for checkboxes and radios, but not set for the `option` element.

Above all that, this behavior (returning `option`'s content if no `value` is provided) [doesn't seem to be mentioned](https://www.w3.org/TR/html-markup/option.html#option) by w3c, but is still implemented in browsers and is expected.

Anyway, to summarize, `option` element looks unique in the sense that it needs the `value` attribute explicitly set even if it is an empty string.

This PR https://github.com/facebook/react/pull/6228 fixes the issue.

**Comment by troydemonbreun** [comment]
@jimfb Was #1510 the PR you were looking for?

**Comment by syranide** [comment]
https://github.com/facebook/react/pull/1510#issuecomment-194722937 and #6119. However, it should be noted that this bug applies to `input` as well, it's just that you don't really notice it. The problem is due to the new renderer and `HAS_SIDE_EFFECTS` for `value` means that the first time `value` is set it's now set using the `DOMPropertyOperations`-logic instead which means the `HAS_SIDE_EFFECTS` case is applied, where previously it would be set as an attribute the first time regardless and `HAS_SIDE_EFFECTS` was thus not taken into consideration. Hence, this is actually a bug of the new renderer in a sense. If we're going with the solution in #6119 then that fixes this too as properties then goes away entirely and is handled in the wrappers instead.

cc @spicyj

**Comment by everdimension** [comment]
@syranide 
I think we should set on the terms whether stripping out empty `value` attribute is a "bug" in general.

`option` element "needs" it because of specific browser handling, so do `checkbox`es and `radio`s. So in these cases empty value should be left, because they might be user inteded.

In other cases though, as I understand it, only attributes classified as `HAS_BOOLEAN_VALUE` should be left. Other attributes, if empty, should be stripped.

So we either handle the `value` attribute as a special case, or we handle the `option` element as a special case.

**Comment by syranide** [comment]
@everdimension There's semantic difference between `<div title="">` and `<div>` even though there's no practical difference, the same applies to inputs. I don't really see a reason to *special-case* `value` for `input` here, but that's my opinion and I don't really mind either. Regardless, the problem and final solution is the same (#6119), if `ReactDOMInput` wants to special-case it then that seems fine to me. Anyway, I leave it up to @spicyj or whoever.

**Comment by chicoxyzzy** [comment]
yep `value=""` have no any reasons to be stripped in any case

**Comment by chicoxyzzy** [comment]
One can put `value=""` into any tag and that could be intended (but really weird tho)

**Comment by jimfb** [comment]
FWIW, if you mount into something that was server side rendered, you get the `value=""` as desired:
```
document.getElementById("container").innerHTML = (ReactDOMServer.renderToString(Select));
ReactDOM.render(Select, document.getElementById("container"));
```
I think a potential solution here would be to stop treating `value` as a special property, and just set it as an attribute, which means this could also be solved by https://github.com/facebook/react/pull/5680 (I think).

**Comment by Chyzwar** [comment]
value="" is permitted only for handful of elements according to HTML spec, you can use data-* cor custom attributes. https://developer.mozilla.org/en/docs/Web/Guide/HTML/Using_data_attributes
